**Using this template**

1. Change notebook title to assignment title you working on, eg. `Graded Challenge 1 or Milestone 1`
2. Put your identity such as name and batch number below notebook title.
3. Write your description before do any work. State your purpose of this assignment and what you are trying to achieve. 
4. Look at the rubric of this assignment and make sure you understand the criteria.
5. Code your solution in the cell provided below the working area section.
6. State your conlusions, findings, and any other relevant information in the cell provided below the conclusions sections.
7. Save this notebook and rename it to assignment title you working on  eg. `Graded Challenge 1 or Milestone 1`
8. Push your assignment before deadline.
9. Good Luck!

# Notebook Title

Live Code 1

Name: Gusti Ayu Dewi Puspa Kartikasari

Batch: 012

## Description

### Objective
Mencari point dan mengerjakan step yang disebutkan dalam problem statement.

### Problem Statement

a. Buatlah dua fungsi eksplisit masing-masing memuat perhitungan hambatan (R) dan daya (P) yang nantinya akan digunakan untuk perhitungan

b. Buat kolom baru dengan nama kolom R/P berisikan hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, Kapasitas <2.5 hitung nilai hambatan. (Hint: Gunakan for loop dan if statement untuk mengerjakan perintah nomor b, kamu bisa menggunakan list untuk menyimpan perhitungan sementara, setelahnya dapat dimasukkan ke kolom R/P).

c. Hitunglah rata-rata nilai kapasitas untuk arus = 1 Ampere dan cycle ganjil.

## Working Area

```txt
Put your code, analysis, everything below this line

Make sure to check everything the rubric requires before you submit
```

#import libraries

In [ ]:
import pandas as pd
import numpy as np

: 

#Import data

In [2]:
df = pd.read_excel('NMC1-9.xlsx')
df.head()

time(s)    V_m  Current (A)  Cycle       Cap
0        0  4.098        0.997      1  2.840817
1        1  4.096        0.998      1  2.840817
2        2  4.094        0.999      1  2.840817
3        3  4.093        0.998      1  2.840817
4        4  4.092        0.999      1  2.840817

Data berhasil di import dengan keterangan kolom:
time(s): Detik ke- pengukuran
V_m: Voltage/tegangan yang terukur tiap waktu
Current (A): Kuat arus terukur tiap waktu
Cycle: Nomor siklus charge-discharge baterai
Cap: Kapasitas baterai tiap waktu -----> yang akan menjadi parameter acuan perhitungan R/P

#a. make 2 function for R and P

In [3]:
def Hambatan_R(Volt_V,Ampere_A):
    return(Volt_V/Ampere_A)

def Daya_P(Volt_V,Ampere_A):
    return(Volt_V*Ampere_A)

#b. Buat kolom baru dengan nama kolom R/P berisikan hasil perhitungan daya/hambatan dengan kriteria: 

Kapasitas >=2.5 hitung daya

Kapasitas <2.5 hitung nilai hambatan. 

Membuat kolom baru untuk hasil tambahan + tambahankan keterangan parameter yang dihitung (untuk kepastian saya saja)

In [4]:
Kolom_hasil = []
Kolom_parameter =[]

for idx in range(0,len(df)):
    
    if df.iloc[idx,4] >= 2.5: #Kapasitas >=2.5 hitung daya
        hasil = Daya_P(df.iloc[idx,1],df.iloc[idx,2])
        Kolom_parameter.append('Daya')

    elif df.iloc[idx,4] < 2.5: #Kapasitas <2.5 hitung nilai hambatan
        hasil = Hambatan_R(df.iloc[idx,1],df.iloc[idx,2])
        Kolom_parameter.append('Hambatan')


    else:
        hasil = 'NaN' #Apabila ada error maka hasil adalah 'NaN'
    
    Kolom_hasil.append(hasil)

df['R/P'] = Kolom_hasil #menambahkan kolom R/P
df['Parameter yang dihitung'] = Kolom_parameter #menambahkan kolom Parameter yg dihitung sebagai acuan
df.head()

time(s)    V_m  Current (A)  Cycle       Cap       R/P  \
0        0  4.098        0.997      1  2.840817  4.085706   
1        1  4.096        0.998      1  2.840817  4.087808   
2        2  4.094        0.999      1  2.840817  4.089906   
3        3  4.093        0.998      1  2.840817  4.084814   
4        4  4.092        0.999      1  2.840817  4.087908   

  Parameter yang dihitung  
0                    Daya  
1                    Daya  
2                    Daya  
3                    Daya  
4                    Daya

Note: Cap: Kapasitas baterai tiap waktu -----> yang akan menjadi parameter acuan perhitungan R/P
semua nilai Kapasitas (Cap) diatas 2.5 ( >= 2.5 ) sehingga semua parameter yang dihitung merupakan Daya
dibuktikan oleh:

In [5]:
df.sort_values(by='Cap', ascending=True).head() #sortir secara ascending untuk melihat nilai Cap terkecil

time(s)    V_m  Current (A)  Cycle      Cap       R/P  \
90981    90981  2.497        1.002      9  2.77018  2.501994   
84331    84331  3.794        1.002      9  2.77018  3.801588   
84330    84330  3.794        1.002      9  2.77018  3.801588   
84329    84329  3.794        1.002      9  2.77018  3.801588   
84328    84328  3.795        1.001      9  2.77018  3.798795   

      Parameter yang dihitung  
90981                    Daya  
84331                    Daya  
84330                    Daya  
84329                    Daya  
84328                    Daya

Nilai terkecil kolom 'Cap' yaitu 2.77018 dimana 2.77018 > 2.5 sehingga semua parameter yang dihitung merupakan Daya

#c. Hitunglah rata-rata nilai kapasitas untuk arus = 1 Ampere dan cycle ganjil.

Rename Current (A) to Current_A agar bisa menggunakan query

In [6]:
df.rename(columns={'Current (A)':'Current_A'}, inplace = True)
df.head()

time(s)    V_m  Current_A  Cycle       Cap       R/P  \
0        0  4.098      0.997      1  2.840817  4.085706   
1        1  4.096      0.998      1  2.840817  4.087808   
2        2  4.094      0.999      1  2.840817  4.089906   
3        3  4.093      0.998      1  2.840817  4.084814   
4        4  4.092      0.999      1  2.840817  4.087908   

  Parameter yang dihitung  
0                    Daya  
1                    Daya  
2                    Daya  
3                    Daya  
4                    Daya

Column renamed, continue in making querry with constrain:
Current_A == 1 and the cycle is odd (cycle %2 != 0)

In [7]:
df.query("Current_A == 1 & \
Cycle % 2 != 0 "
)


time(s)    V_m  Current_A  Cycle       Cap    R/P  \
15          15  4.085        1.0      1  2.840817  4.085   
19          19  4.083        1.0      1  2.840817  4.083   
20          20  4.083        1.0      1  2.840817  4.083   
21          21  4.082        1.0      1  2.840817  4.082   
31          31  4.079        1.0      1  2.840817  4.079   
32          32  4.078        1.0      1  2.840817  4.078   
36          36  4.077        1.0      1  2.840817  4.077   
46          46  4.074        1.0      1  2.840817  4.074   
67          67  4.068        1.0      1  2.840817  4.068   
68          68  4.068        1.0      1  2.840817  4.068   
78          78  4.066        1.0      1  2.840817  4.066   
81          81  4.065        1.0      1  2.840817  4.065   
82          82  4.065        1.0      1  2.840817  4.065   
91          91  4.063        1.0      1  2.840817  4.063   
102        102  4.061        1.0      1  2.840817  4.061   
107        107  4.060        1.0      1  2.840817  4.060   
113        113  4.059        1.0      1  2.840817  4.059   
132        132  4.055        1.0      1  2.840817  4.055   
135        135  4.055        1.0      1  2.840817  4.055   
141        141  4.054        1.0      1  2.840817  4.054   
143        143  4.054        1.0      1  2.840817  4.054   
149        149  4.053        1.0      1  2.840817  4.053   
152        152  4.052        1.0      1  2.840817  4.052   
161        161  4.051        1.0      1  2.840817  4.051   
167        167  4.050        1.0      1  2.840817  4.050   
170        170  4.049        1.0      1  2.840817  4.049   
176        176  4.049        1.0      1  2.840817  4.049   
180        180  4.048        1.0      1  2.840817  4.048   
186        186  4.047        1.0      1  2.840817  4.047   
190        190  4.046        1.0      1  2.840817  4.046   
...        ...    ...        ...    ...       ...    ...   
90693    90693  2.887        1.0      9  2.770180  2.887   
90706    90706  2.877        1.0      9  2.770180  2.877   
90721    90721  2.866        1.0      9  2.770180  2.866   
90722    90722  2.865        1.0      9  2.770180  2.865   
90723    90723  2.865        1.0      9  2.770180  2.865   
90752    90752  2.841        1.0      9  2.770180  2.841   
90753    90753  2.840        1.0      9  2.770180  2.840   
90760    90760  2.834        1.0      9  2.770180  2.834   
90767    90767  2.828        1.0      9  2.770180  2.828   
90768    90768  2.827        1.0      9  2.770180  2.827   
90781    90781  2.815        1.0      9  2.770180  2.815   
90782    90782  2.813        1.0      9  2.770180  2.813   
90797    90797  2.798        1.0      9  2.770180  2.798   
90806    90806  2.789        1.0      9  2.770180  2.789   
90815    90815  2.778        1.0      9  2.770180  2.778   
90819    90819  2.773        1.0      9  2.770180  2.773   
90827    90827  2.764        1.0      9  2.770180  2.764   
90831    90831  2.759        1.0      9  2.770180  2.759   
90840    90840  2.747        1.0      9  2.770180  2.747   
90854    90854  2.728        1.0      9  2.770180  2.728   
90859    90859  2.721        1.0      9  2.770180  2.721   
90868    90868  2.707        1.0      9  2.770180  2.707   
90877    90877  2.693        1.0      9  2.770180  2.693   
90893    90893  2.668        1.0      9  2.770180  2.668   
90903    90903  2.651        1.0      9  2.770180  2.651   
90909    90909  2.640        1.0      9  2.770180  2.640   
90923    90923  2.615        1.0      9  2.770180  2.615   
90937    90937  2.589        1.0      9  2.770180  2.589   
90949    90949  2.565        1.0      9  2.770180  2.565   
90967    90967  2.528        1.0      9  2.770180  2.528   

      Parameter yang dihitung  
15                       Daya  
19                       Daya  
20                       Daya  
21                       Daya  
31                       Daya  
32                       Daya  
36                       Daya  
46                       Daya 

Recheck data if all odd cycle already included in the query

In [8]:
df.query("Current_A == 1 & \
Cycle % 2 != 0 "
).groupby(by = ['Current_A','Cycle'])[['Cap']].mean()

Cap
Current_A Cycle          
1.0       1      2.840817
          3      2.823520
          5      2.804878
          7      2.786860
          9      2.770180

All data from cycle 1,3,5,7,9 (all odd) has been included. Proceed to calculate mean of capacity (Cap).

In [9]:
df.query("Current_A == 1 & \
Cycle % 2 != 0 "
)[['Cap']].mean()

Cap    2.806485
dtype: float64

The mean of capacity with constraints as follows: Current_A == 1 and the cycle is odd (cycle %2 != 0)
is 2.806485

## Conclusions, Assumptions, Overall Analysis

a. Fuctions are made under function of: Hambatan_R and Daya_P

b. New column are made for R/P using for loop and if. From here can be noted that the minimum capacity is still bigger than 2.5. Therefore, all of parameter that filled R/P column are Daya_P (Watt).

c. The mean of capacity with constraints as follows: Current_A == 1 and the cycle is odd (cycle %2 != 0)
is 2.806485